### ISSUE TITLE
Tagging of Volume

### ISSUE THREAD

**Issue by demalik** [NetApp/harvest#209]
Customer would like to have capability to tag volumes for business units, workloads, projects, cost center etc etc.

In customer environment volume names are using artificial, random names during the provisioning process. For proper identification they would like tags for project, cost-center etc

Q: Can you please advise us on how to add custom tags to time series for volumes. Like BusinessUnit, Project, UsageModel etc…  
A: Prometheus supports labels (which can be treated like tags). A counter can have many labels.
Typically we assign labels to counters like:
•	datacenter
•	cluster
•	vserver
•	volume
•	…

Labels get assigned during ingest into the database. You cannot (like with all TSDBs) add tags
or labels afterwards. If you want to add custom labels to volumes as outlined below, there
are actually three options to do that:
1.	you use the comment filed in ONTAP itself – you need to create a plugin in Harvest
to decode the comment field and create labels for the volume. We would recommend
to create a new counter per volume which holds all these labels  - this is more efficient
and can be used for filtering in Grafana
2.	You load this new volume counter manually into Prometheus using the file exporter
capability of Prometheus node exporter
3.	You create a new collector which generates these counters

It’s more the question who and when do you want to assign these tags to the volume

For option 1: Can the netapp team create that plugin… I am sure this is something that all your customer base will find useful.  

Q: Sorry if this is a repeated question, but can you provide 1-2 examples of the comment section and the labels you want to extract?
A: How about the following

Comment= “ Label1:<label data> Label2:<label data> Label 3:<label data>  Label4:<label data> ………”
Example Comment=” Business_group:XYZ  project:my_project  user1:myuser1 user2:my_user2 … and many more”

Each label:label data separated by spaces.. You can interpret as many labels or set a limit of 10 labels.   

Label data and label names cannot have any special characters other than “_ or – “ and alphanumeric characters only

We can start with volume management. But if easy to implement for other objects we should. But volume is priority

Q: If the Zapi object does provide the comment as a metric, then it's easy and probably no need to write a new plugin for another object.
One last question that might sound a bit strange, but makes a big difference how the plugin would be implemented: how are you planning to use these labels? Do you want to be able to filter/select the metrics based on these labels? Or is it sufficient for you to get a table showing these labels (e.g. which volumes have which owners, projects, etc.). 

A: We expect to filter based on these metrics and do various operation. Also we plan to implement alerting and monitoring also based on these filters. 

Example: If utilization >90 for businessunit:xyz  then send email or raise alert  etc…

**Comment by ruanruijuan** [comment]
Reference of the volume "comment" field:
http://web.eng.netapp.com/engineering/rlse/DOT/devN/final/bedrock/export/common/zapiprod/web_ontap_cluster/volume/index.html#volume-attributes and search for "comment" (internal to netapp). It has a restapi mapping.

**Comment by vgratian** [comment]
> Comment= “ Label1: Label2: Label 3: Label4: ………”
>Example Comment=” Business_group:XYZ project:my_project user1:myuser1 user2:my_user2 … and many more”
>
>Each label:label data separated by spaces.. You can interpret as many labels or set a limit of 10 labels.

@damalik, sorry I lost track of our conversation about this issue, but this should not require any new feature. Splitting key-value pairs from a label is possible with the built-in plugin [LabelAgent](https://github.com/NetApp/harvest/tree/main/cmd/poller/plugin#labelagent). All you have to do is edit the templates of the Zapi/ZapiPerf objects where you are inserting this comment label. 

Assuming that is the case for the Volume object, you would need to change the [volume.yaml](https://github.com/NetApp/harvest/blob/main/conf/zapiperf/cdot/9.8.0/volume.yaml) template. There are two sections that you need to change.

**1.** - In the `plugins` section, add a rule for the LabelAgent plugin. The following should work for the example you brought above:

in the original template, the plugins section looks something like this:
```yaml
plugins:
  - Volume
```

change that to:
```yaml
plugins:
  Volume
  LabelAgent:
    split_pairs: comment ` ` `:`
```
**2.** - change the `export_options` section to make sure the extracted labels are included in the final metric. Again, I take here your example:

the section should originally look something like this:

```yaml
export_options:
  instance_keys:
    - volume
    - node
    - svm
    - aggr
    - type
```
you should change it to look something like this:

```yaml
export_options:
  instance_keys:
    - volume
    - node
    - svm
    - aggr
    - type
    - Business_group
    - project
    - user1
    - user2
```

**Comment by vgratian** [comment]
if the set of labels that need to be extracted from `comment` are not pre-defined, you can simply change the export_options section to:

```yaml
export_options:
  include_all_labels: true
```

**Comment by cgrinds** [comment]
We'll use this issue to update the documentation and validate this. Thanks @vgratian

**Comment by demalik** [comment]
> if the set of labels that need to be extracted from `comment` are not pre-defined, you can simply change the export_options section to:
> 
> ```yaml
> export_options:
>   include_all_labels: true
> ```

**Comment by demalik** [comment]
Here is feedback from customer, something does not seems to be working

Hi Deepak
            I have update volume.yaml file as suggested below.

 


Updated the comment for a volume as below.

xyz-clu01::> vol modify macharya_test -vserver xyz-clu01-scr04 -comment "business_unit:IT project:unix workflows:test owner:macharya bu_project:IT"
Volume modify successful on volume macharya_test of Vserver xyz-clu01-scr04.

xyz-clu01::> volume show macharya_test -vserver xyz-clu01-scr04 -fields comment
vserver         volume        comment
--------------- ------------- -------------------------------------------------------------------------
xyz-clu01-scr04 macharya_test business_unit:IT project:unix workflows:test owner:macharya bu_project:IT

Restarted the poller after this and searched for this volume in Prometheus. 

 

Still don’t see comments added displayed as labels. How long we need to wait for the labels to show up? Is there any typo in my volume.yaml file?

**Comment by vgratian** [comment]
Hi @demalik, I see that there were a few mistakes in my suggestion, below are the steps once again, this time they are correct.

Main mistake was to suggest to use the ZapiPerf collector (it does not provide the comment section as a counter), so instead please change the [volume.yaml](conf/zapi/cdot/9.8.0/volume.yaml) template of the Zapi collector:

(1) First make sure that "comment" is collected, add it under the `counters` section, under the `volume-id-attributes`:

```yaml
counters:
  volume-attributes:
    - volume-autosize-attributes:
      - maximum-size
      - grow-threshold-percent

    - volume-id-attributes:
      - ^^instance-uuid             => instance_uuid
      - ^name                       => volume
      - ^node                       => node
      - ^owning-vserver-name        => svm
      - ^containing-aggregate-name  => aggr
      - ^style-extended             => style
      - ^comment                    => comment  # only this line needs to be added
     ....
```

(2) Under the `plugins` section add the exact same split_pairs rule, the section will look like this:

```yaml
plugins:
  LabelAgent:
    value_mapping: status state online `1`
    exclude_equals: style `flexgroup_constituent`
    replace: style style `flexgroup_constituent` `flexgroup`
    split_pairs: comment ` ` `:`  # only this line is added
  Aggregator:
    volume<style=flexgroup>volume node,svm,aggr,style

```

(3) Finally under `export_options`, add the labels that you expected to be extracted, e.g.:

```yaml
export_options:
  instance_keys:
    - volume
    - node
    - svm
    - aggr
    - style
    - business_unit
    - project
    - owner
    - bu_project
  instance_labels:
    - state
```

Here is how I tested it:

- Added comment:
```bash
# Using the ONTAP CLI:
shopfloor::> vol modify -volume vgratian_tmp_will_delete -vserver svm0 -comment "business_unit:IT project:unix workflows:test owner:macharya bu_project:IT" 
Volume modify successful on volume vgratian_tmp_will_delete of Vserver svm0.                                                                                                                 

```

- Started Harvest:
```bash
harvest start shopfloor -f --config harvest_.yml -c Zapi -o Volume
```

- Tested the metrics:

```bash
sanjunipero>$ curl localhost:5551/metrics | grep vgratian_
volume_labels{datacenter="WDRF",cluster="shopfloor",volume="vgratian_tmp_will_delete",node="shopfloor-02",svm="svm0",aggr="aggr02",style="flexvol",business_unit="IT",project="unix",owner="macharya",bu_project="IT",state="online"} 1.0
volume_size{datacenter="WDRF",cluster="shopfloor",volume="vgratian_tmp_will_delete",node="shopfloor-02",svm="svm0",aggr="aggr02",style="flexvol",business_unit="IT",project="unix",owner="macharya",bu_project="IT"} 1073741824
volume_size_total{datacenter="WDRF",cluster="shopfloor",volume="vgratian_tmp_will_delete",node="shopfloor-02",svm="svm0",aggr="aggr02",style="flexvol",business_unit="IT",project="unix",owner="macharya",bu_project="IT"} 1020055552
....
```

**Comment by vgratian** [comment]
I will take a look for the case when the extracted label set are not predefined, not sure if we can do that with the currently available plugins..

**Comment by vgratian** [comment]
`include_all_labels: true` should work, but it will add a lot of redundant labels, including the original `comment` label.

**Comment by vgratian** [comment]
@Hardikl, sorry, I hope you didn't already spend time on this, felt obliged to correct my mistakes

**Comment by Hardikl** [comment]
@vgratian I haven't started looking into this yet.

**Comment by cgrinds** [comment]
Closed as part of #309. Feel free to reopen @madhusudhanarya if there's more to cover here. I added this information to the [wiki](https://github.com/NetApp/harvest/wiki/FAQ#tagging-volumes).
